### ISSUE TITLE
is that possible to change default image from BGR to RGB at 5.0 version?

### ISSUE THREAD

**Issue by zihaomu** [opencv/opencv#23906]
### Describe the feature and motivation

Users always do the following:

```.py
img_bgr = cv2.imread("PATH") # OpenCV decode the image and swap B and R.
im_rgb = cv2.cvtColor(img_bgr, cv2.COLOR_BGR2RGB) # than swap R and B back.
```

**Comment by dkurt** [comment]
That will add incompatibility between OpenCV 5 and older code.

I'd rather expect adding a flag for [imread](https://docs.opencv.org/4.x/d8/d6a/group__imgcodecs__flags.html#ga61d9b0126a3e57d9277ac48327799c80): `IMREAD_COLOR_RGB`.

**Comment by vpisarev** [comment]
@zihaomu, yes, we already discussed it a bit, I think. I agree with @dkurt that just changing BGR to RGB will break a **lot** of code. We need some mechanism to provide smooth transition. One of the possible solution is to add a `colorspace` field to `cv::Mat` (or maybe `layout`). By default it could be `0` or `COLORSPACE_UNKNOWN/LAYOUT_UNKNOWN`. Most of the image processing functions (resize, warping, filters) do not change colorspace, so they can copy it through. `imread` could return RGB image if user asks so, as @dkurt suggested. Then, we may conditionally swap channels during preprocessing before inference if the model requires certain channel order. Such a method would be more robust

**Comment by asmorkalov** [comment]
IMHO `IMREAD_COLOR_RGB` is enough as first step to give advantage to migrate to 5.x right after release.

**Comment by zihaomu** [comment]
Got it. I can be assigned this task if needed.

**Comment by ocpalo** [comment]
The PR from my GSoC proposal is still open and needs some work to implement it. It was based on Vadim`s [named parameters](https://github.com/opencv/opencv/wiki/OE-34.-Named-Parameters) suggestion. 

If you discuss that kind of approach and finalize your decision, I can give it a shot to finalize this. So adding this kinds of request would not break(atleast not API, not sure about ABI). This flag can be added to that new `imread` method that I created.

Pullrequest is : https://github.com/opencv/opencv/pull/22431

**Comment by atinfinity** [comment]
> One of the possible solution is to add a `colorspace` field to `cv::Mat` (or maybe `layout`). By default it could be `0` or `COLORSPACE_UNKNOWN/LAYOUT_UNKNOWN`. 

This is very interesting topic. And, I have some concern.

### concern1

User can create instance from image buffer.

```c++
cv::Mat img = cv::Mat(rows, cols, CV_8UC3, buffer);
```

In this case, do you assume that `COLORSPACE_UNKNOWN` is set to `colorspace`?

### concern2

When users use OpenCV python binding, users may use different framework to read image/video.
I concerned that `colorspace` is not set in this case.

### concern3

In OpenCV python binding, users use data as `numpy.ndarray`.
How is colourspace handled?

**Comment by cuijianaaa** [comment]
Any update?

**Comment by zihaomu** [comment]
@cuijianaaa, Try to create PR next week.

**Comment by Avasam** [comment]
#25809 Added a flag for RGB, which is nice. But unhelpful for images with alpha channels, `IMREAD_UNCHANGED` is still the only flag that keeps transparency according to doc.

**Comment by themightyoarfish** [comment]
It would also be good to have a flag for imwrite(), to avoid producing bgr images on disk

**Comment by Chiramisudo** [comment]
> That will add incompatibility between OpenCV 5 and older code.

There are already a ton of other breaking changes anyway, and the fact that the default was ever BGR to begin with is, in my mind, bonkers. 🤦🏼‍♂️

Also, as a matter of good practice when writing code, **ALWAYS** be explicit and never rely on defaults! You're just asking for pain when you do that.
